server/client.go

```diff
@@ -2205,11 +2254,31 @@ func (c *client) pubAllowed(subject string) bool {
 		r := c.perms.pub.deny.Match(subject)
 		allowed = len(r.psubs) == 0
 	}
-	// Update our cache here.
-	c.perms.pcache[string(subject)] = allowed
-	// Prune if needed.
-	if len(c.perms.pcache) > maxPermCacheSize {
-		c.prunePubPermsCache()
+
+	// If we are currently not allowed but we are tracking reply subjects
+	// dynamically, check to see if we are allowed here Avoid pcache.
+	// We need to acquire the lock though.
+	if !allowed && c.perms.resp != nil {
+		c.mu.Lock()
```

Comment:
That is going to be a problem. `pubAllowed` is called by `canImport` that is holding the connection's lock. Let's check that.

Reply:
ok will take a closer look.

Reply:
But then this means that when invoked from a route, that won't work. That is, if we allow these kind of permissions for routes, then they won't be checked.

Reply:
instead, I would have conditionally lock/unlock since in one case we already have the lock, other we don't.

Reply:
I disallowed from cluster config.